docs/overview.rst

```diff
@@ -1,11 +1,324 @@
 Overview
 ========
 
-The data pipeline abstraction that Fuel provides looks complicated at first
-sight, but can be very powerful.
+We'll go over a quick example to see what Fuel is capable of.
+
+Let's start by creating some random data to act as features and targets. We'll
+pretend that we have eight 2x2 grayscale images separated into four classes.
+
+>>> import numpy
+>>> seed = 1234
+>>> rng = numpy.random.RandomState(seed)
+>>> features = rng.randint(256, size=(8, 2, 2))
+>>> targets = rng.randint(4, size=(8, 1))
+
+Our goal is to use Fuel to interface with this data, iterate over it in various
+ways and apply transformations to it on the fly.
+
+Division of labour
+------------------
+
+There are four basic tasks that Fuel needs to handle:
+
+* Interface with the data, be it on disk or in memory.
+* Decide which data points to visit, and in which order.
+* Iterate over the selected data points.
+* At each iteration step, apply some transformation to the selected data points.
+
+Each of those four tasks is delegated to a particular class of objects, which
+we'll be introducing in order.
+
+Datasets: interfacing with data
+-------------------------------
+
+The :class:`Dataset` class is responsible for interfacing with the data and
+handling data access requests. Subclasses of :class:`Dataset` specialize in
+certain types of data.
+
```

Comment:
Datasets contain one or more **sources** of data such as, e.g., an array of images, a list of labels, a dictionary specifying an ontology, etc, ... Each source in a Dataset is identified by a unique name.